core/primitives/debug-derive/src/lib.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Macros to derive runtime debug implementation.
+
+extern crate proc_macro;
+
+mod impls;
+
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(RuntimeDebug)]
+pub fn debug_derive(input: TokenStream) -> TokenStream {
+	match syn::parse(input) {
```

Comment:
Why not use `parse_macro_input!`?

Reply:
We pretty much handle all structs and enums anyway. I think the custom type would pretty much look exactly like the generic ones from `syn`.
Unless I'm missing some obvious advantage.

Reply:
```
impls::debug_derive(&parse_macro_input!(input))
```

This would be all you for the function. It automatically returns an error internally, if the parsing failes.